components/sync15/src/telemetry.rs

```diff
@@ -0,0 +1,645 @@
+#[derive(Debug, Serialize)]
+struct WhenTook {
+    when: f64,
+    #[serde(skip_serializing_if = "skip_if_zero")]
+    took: u64,
+}
+
+/// What we track while recording 'when' and 'took. It serializes as a WhenTook,
+/// except when .finished() hasn't been called, in which case it panics.
+#[derive(Debug)]
+enum Stopwatch {
+    Started(time::SystemTime, time::Instant),
+    Finished(WhenTook),
+}
+
+impl Stopwatch {
+    fn new() -> Self {
+        Stopwatch::Started(time::SystemTime::now(), time::Instant::now())
+    }
+
+    fn finished(&self) -> Self {
+        match cfg!(test) {
+            false => self._finished(),
+            _ => Stopwatch::Finished(WhenTook { when: 0.0, took: 0 }),
+        }
+    }
+
+    fn _finished(&self) -> Self {
+        match self {
+            Stopwatch::Started(st, si) => {
+                let std = st
+                    .duration_since(time::UNIX_EPOCH)
+                    .unwrap_or(time::Duration::new(0, 0));
+                let when = std.as_secs() as f64; // we don't want sub-sec accuracy. Do we need to write a float?
+
+                let sid = si.elapsed();
+                let took = sid.as_secs() * 1000 + (sid.subsec_nanos() as u64) / 1_000_000;
+                Stopwatch::Finished(WhenTook { when, took })
+            }
+            _ => {
+                panic!("can't finish twice");
+            }
+        }
+    }
+}
+
+impl Serialize for Stopwatch {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Stopwatch::Started(_, _) => panic!("must be finished"),
+            Stopwatch::Finished(c) => c.serialize(serializer),
+        }
+    }
+}
+
+#[cfg(test)]
+mod stopwatch_tests {
+    use super::*;
+
+    // A wrapper struct because we flatten - this struct should serialize with
+    // 'when' and 'took' keys (but with no 'sw'.)
+    #[derive(Debug, Serialize)]
+    struct WT {
+        #[serde(flatten)]
+        sw: Stopwatch,
+    }
+
+    #[should_panic]
+    #[test]
+    fn test_not_finished() {
+        let wt = WT {
+            sw: Stopwatch::new(),
+        };
+        serde_json::to_string(&wt).expect("a panic!");
+    }
+
+    #[test]
+    fn test() {
+        assert_json(
+            &WT {
+                sw: Stopwatch::Finished(WhenTook { when: 1.0, took: 1 }),
+            },
+            json!({"when": 1.0, "took": 1}),
+        );
+        assert_json(
+            &WT {
+                sw: Stopwatch::Finished(WhenTook { when: 1.0, took: 0 }),
+            },
+            json!({"when": 1.0}),
+        );
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+//
+// A generic "Event" - suitable for all kinds of pings!
+// (although currently we only support the sync ping...)
+//
+#[derive(Debug, Serialize)]
+pub struct Event {
+    // We use static str references as we expect values to be literals.
```

Comment:
I'm happy about this!

It wouldn't surprise me to see it go eventually (the values might still be literals, just on the other side of the FFI), but lets keep it for now.